tfhe/src/core_crypto/algorithms/lwe_multi_bit_programmable_bootstrapping.rs

```diff
@@ -6,28 +6,87 @@ use crate::core_crypto::commons::math::decomposition::SignedDecomposer;
 use crate::core_crypto::commons::parameters::*;
 use crate::core_crypto::commons::traits::*;
 use crate::core_crypto::entities::*;
-use crate::core_crypto::fft_impl::common::pbs_modulus_switch;
+use crate::core_crypto::fft_impl::common::modulus_switch;
 use crate::core_crypto::fft_impl::fft64::crypto::ggsw::{
     add_external_product_assign, add_external_product_assign_scratch, update_with_fmadd_factor,
 };
 use crate::core_crypto::fft_impl::fft64::math::fft::{Fft, FftView};
+use aligned_vec::ABox;
 use concrete_fft::c64;
+use itertools::Itertools;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::{mpsc, Condvar, Mutex};
 use std::thread;
 
-pub fn prepare_multi_bit_ggsw_mem_optimized<
-    Scalar,
-    GgswBufferCont,
-    GgswGroupCont,
-    FourierPolyCont,
->(
+pub(crate) fn modulus_switch_multi_bit<Scalar>(
+    ciphertext_modulus_log: CiphertextModulusLog,
+    grouping_factor: LweBskGroupingFactor,
+    lwe_mask_elements: &[Scalar],
+) -> impl Iterator<Item = usize> + '_
+where
+    Scalar: UnsignedInteger + CastInto<usize> + CastFrom<usize>,
+{
+    // Start at 1, the first ggsw is not rotated
+    (1..grouping_factor.ggsw_per_multi_bit_element().0).map(move |ggsw_idx| {
+        // Select the proper mask elements to build the monomial degree depending on
+        // the order the GGSW were generated in, using the bits from mask_idx and
+        // ggsw_idx as selector bits
+        let mut monomial_degree = Scalar::ZERO;
+        for (mask_idx, &mask_element) in lwe_mask_elements.iter().enumerate() {
+            let mask_position = lwe_mask_elements.len() - (mask_idx + 1);
+            let selection_bit: Scalar = Scalar::cast_from((ggsw_idx >> mask_position) & 1);
+            monomial_degree =
+                monomial_degree.wrapping_add(selection_bit.wrapping_mul(mask_element));
+        }
+        modulus_switch(monomial_degree, ciphertext_modulus_log).cast_into()
+    })
+}
+
+pub trait MultiBitmodulusSwitchedCt: Sync {
+    fn switched_modulus_input_lwe_body(&self) -> usize;
+    fn switched_modulus_input_mask_per_group(
+        &self,
+        index: usize,
+    ) -> impl Iterator<Item = usize> + '_;
+}
+
+pub struct StandardMultiBitmodulusSwitchedCt<
+    'a,
+    Scalar: UnsignedInteger + CastInto<usize> + CastFrom<usize>,
+> {
+    input: &'a LweCiphertext<&'a [Scalar]>,
+    grouping_factor: LweBskGroupingFactor,
+    log_modulus: CiphertextModulusLog,
+}
```

Comment:
this is public so please use the Container approach here

Reply:
This object is not meant to be stored, it's a kind of iterator
I don't think it's worth the effort

Reply:
It's not about storing, it's about usability without having to call .as_view

Reply:
Done